### PR TITLE
BigtableVersion[Range] wrapper types

### PR DESCRIPTION
**Pull request by evildour**

### CODE REVIEW

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersionRange.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        public override bool Equals(object obj) => Equals(obj as BigtableVersionRange);
+
+        /// <inheritdoc />
+        public override int GetHashCode() =>
+            Start.HasValue.GetHashCode() ^ End.HasValue.GetHashCode() ^
+            Start.GetHashCode() ^ End.GetHashCode();
+
+        /// <inheritdoc />
+        public override string ToString() =>
+            $"{nameof(BigtableVersionRange)}: Start={Start?.Value.ToString() ?? " "}; End={End?.Value.ToString() ?? " "}";
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersionRange.cs]: Consider "(none)" or "<>" as an alternative to " " for a missing value?

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            Value = timestamp.Ticks / TicksPerMicro;
+        }
+
+        /// <summary>
+        /// Gets the version value. Greater version values indicate newer cell values.
+        /// </summary>
+        public long Value { get; }
+
+        /// <summary>
+        /// Gets the DateTime equivalent to the version assuming the value is a timestamp micros value.
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: I see we're using the .NET epoch (0001-01-01T00:00:00). Is that what other languages are likely to do? Might the Unix epoch work better? (Just thinking about interop really.)

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            }
+
+            return x.Value.CompareTo(y.Value);
+        }
+
+        /// <inheritdoc />
+        public bool Equals(BigtableVersion other) => CompareTo(other) == 0;
+
+        /// <inheritdoc />
+        public override bool Equals(object obj) => obj is BigtableVersion other && Equals(other);
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Just `Equals(other as BigtableVersion)` given the null result works? (No particular need either way.)

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        /// <remarks>
+        /// <para>
+        /// Note: when using ReadModifyWriteRow, modified columns automatically use a server version, which
+        /// is based on the current timestamp micros. For those columns, other reads and writes should
+        /// use <see cref="BigtableVersion"/> values constructed from DateTime values, as opposed to using
+        /// a custom versioning scheme with 64-bit values.
+        /// </para>
+        /// </remarks>
+        /// <param name="value">
+        /// The non-negative version value, or -1 to initialize from the microseconds of DateTime.UtcNow.
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Just to check, does the Bigtable team recommend using client-side clocks like this?

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionRangeTest.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Xunit;
+
+namespace Google.Cloud.Bigtable.V2.Tests
+{
+    public class BigtableVersionRangeTest
+    {
+        [Fact]
+        public void Equality()
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionRangeTest.cs]: Equality check with a null start/end value as well?

Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionRangeTest.cs]: Done

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs (hunk 4 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Yes, we had a thread about this in the API proposal doc with Gary Elliot.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs (hunk 2 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Oops, yes you're right. I'll fix this (in code and docs)

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs (hunk 3 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: `BigtableVersion` is a struct, so it isn't nullable.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersionRange.cs (hunk 1 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersionRange.cs]: Done (using "(none)")

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Doh - sorry about that :)

**Comment by evildour** [comment]
Thanks Jon. PTAL.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs [hunk 6]
```diff
@@ -67,7 +67,7 @@ public void RoundTripLong(long value)
         }
 
         [Theory]
-        [InlineData("BigtableVersion: 63643449600000000", "2017-10-13Z")]
+        [InlineData("BigtableVersion: 1507852800000000", "2017-10-13Z")]
         public void FormattingDateTime(string expectedText, string dateTimeText)
         {
             var dateTime = DateTime.Parse(dateTimeText).ToUniversalTime();
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs]: Hadn't spotted this before. Better to use `DateTime.ParseExact`, specifying the invariant culture and `DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal`

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs [hunk 7]
```diff
@@ -73,7 +76,7 @@ public BigtableVersion(DateTime timestamp)
                 nameof(timestamp),
                 $"The {nameof(BigtableVersion)} timestamp must be specified in UTC.");
 
-            Value = timestamp.Ticks / TicksPerMicro;
+            Value = (timestamp.Ticks - UnixEpoch.Ticks) / TicksPerMicro;
```
Comment (jskeet) [inline apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: It's not easy to tell, but presumably you've verified that you've caught all the places you use DateTime?

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs (hunk 6 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs]: Thanks. I used "u" as the format, but let me know if there's something better to use.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs (hunk 7 above)
Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableVersion.cs]: Nope, I missed one. I'll add the test that should have broken.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs (hunk 6 above)
Comment (jskeet) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs]: I'd just be explicit: "yyyy-MM-dd" (with the invariant culture) and remove "Z" from the input strings. That's probably simpler for most devs to read than remembering what "u" means.

Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs]: `ParseExact` requires a format string though.

Comment (evildour) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2.Tests/BigtableVersionTest.cs]: Oh, I misunderstood. Sorry

**Comment by evildour** [comment]
PTAL

**Pull request merged**